sw/device/lib/dif/dif_entropy.h

```diff
@@ -0,0 +1,678 @@
+ *
+ * This struct describes runtime information for one-time configuration of the
+ * hardware.
+ */
+typedef struct dif_entropy_config {
+  /**
+   * The mode to configure the entropy source in.
+   */
+  dif_entropy_mode_t mode;
+
+  /**
+   * Which health tests to enable.
+   *
+   * Note that the value at `kDifEntropyTestMailbox` is ignored, since this test
+   * is driven by the firmware, not the hardware.
+   */
+  bool tests[kDifEntropyTestNumVariants];
+
+  /**
+   * If set, all health-test-related registers will be cleared.
+   */
+  bool reset_health_test_registers;
+
+  /**
+   * Specifies which single-bit-mode to use, if any at all.
+   */
+  dif_entropy_single_bit_mode_t single_bit_mode;
+
+  /**
+   * If set, entropy will be routed to a firmware-visible register instead of
+   * being distributed to other hardware IPs.
+   */
+  // Open Question: Make this its own function? Seems like something that would
+  // be toggled a lot.
+  bool route_to_firmware;
+
+  /**
+   * The mode to perform health tests in.
+   *
+   * Among other things, this sets the type of conditioning performed on data,
+   * and which registers the health test parameters below are written to.
+   */
+  // Open Question: Maybe make this and the u16 fields below a separate
+  // configuration??
+  dif_entropy_test_mode_t test_mode;
+
+  /**
+   * Configuration parameters for health tests.
+   */
+  dif_entropy_test_config_t test_config;
+
+  /**
+   * The rate at which the entropy bits are generated, in clock cycles.
+   */
+  uint16_t sample_rate;
+} dif_entropy_config_t;
+
+/**
+ * A handle to an entropy source.
+ *
+ * This type should be treated as opaque by users.
+ */
+typedef struct dif_entropy {
+  dif_entropy_params_t params;
+} dif_entropy_t;
+
+/**
+ * Revision information for an entropy source.
+ *
+ * The fields of this struct have an implementation-specific interpretation.
+ */
+typedef struct dif_entropy_revision {
+  uint8_t abi_revision;
+  uint8_t hw_revision;
+  uint8_t chip_type;
+} dif_entropy_revision_t;
+
+/**
+ * Statistics on entropy source health tests.
+ */
+// Open Question: I'm not excited by this struct, but I really don't want to
+// have one field per datum...
+typedef struct dif_entropy_test_stats {
+  /**
+   * Watermarks indicating where the value emitted by a particular test has
+   * ranged through; the low watermark is the lowest observed value, while the
+   * high watermark is the highest.
+   *
+   * Each pair of watermarks is presented as a range from lowest to highest.
+   */
+  uint16_t watermarks[2][kDifEntropyTestNumVariants];
+
+  /**
+   * The number of times a particular test has failed.
+   *
+   * For tests that ensure the value lies in a range (such as the "Markov"
+   * test), the array will contain the number of underflows and overflows of
+   * this range, respectively; for tests that only have an upper range, the
+   * first array element will be the number of fails, and the second will be
+   * zeroed.
```

Comment:
How about zeroing the first element and setting the second element to be consistent with other entries? My reasoning: tests that don't have a lower range cannot underflow, thus the first element will always be zero.

Also, it may be a good idea to explicitly mention repcnt and bucket.